lib/service/cfg.go

```diff
@@ -541,6 +542,38 @@ type App struct {
 	Rewrite *Rewrite
 }
 
+// Check validates an application.
+func (a App) Check() error {
+	if a.Name == "" {
+		return trace.BadParameter("missing name")
+	}
+	if a.URI == "" {
+		return trace.BadParameter("missing URI")
+	}
+	// Check if the application name is a valid subdomain. Don't allow names that
+	// are invalid subdomains because for trusted clusters the name is used to
+	// construct the domain that the application will be available at.
+	if errs := validation.IsDNS1035Label(a.Name); len(errs) > 0 {
+		return trace.BadParameter("application name %q must be a valid DNS subdomain: https://gravitational.com/teleport/docs/application-access/#application-name", a.Name)
+	}
+	// Parse and validate URL.
+	if _, err := url.Parse(a.URI); err != nil {
+		return trace.Wrap(err)
+	}
+	// If a port was specified or a IP address was provided for the public
+	// address, return an error.
+	if a.PublicAddr != "" {
+		if _, _, err := net.SplitHostPort(a.PublicAddr); err == nil {
+			return trace.BadParameter("public address %q can not contan a port, applications will be available on the same port as the web proxy", a.PublicAddr)
+		}
+		if net.ParseIP(a.PublicAddr) != nil {
+			return trace.BadParameter("public address %q can not be an IP address, Teleport Application Access uses DNS names for routing", a.PublicAddr)
```

Comment:
I think AAP using DNS names for routing is implementation detail that should not matter in the error.